### PR TITLE
Support deploying to static websites

### LINKED ISSUE #146: Add basic support for viewing and deploying to static websites

### PR DESCRIPTION
**Pull request by StephenWeatherford**
Fixes #146

### CODE REVIEW

File: src/azureStorageExplorer/blobContainers/blobContainerNode.ts [hunk 1]
```diff
@@ -76,6 +79,25 @@ export class BlobContainerNode implements IAzureParentTreeItem, ICopyUrl {
...
+        let currentToken: any;
+        let blobs: azureStorage.BlobService.BlobResult[] = [];
+
+        // tslint:disable-next-line:no-constant-condition
+        while (true) {
+            this.throwIfCanceled(cancellationToken, properties, "listAllBlobs");
+            let result = await this.listBlobs(currentToken, 5000);
+            blobs.push(...result.entries);
+            currentToken = result.continuationToken;
+            if (!currentToken) {
```
Comment (nturinski) [inline src/azureStorageExplorer/blobContainers/blobContainerNode.ts]: I'm assuming that at some point, result.continuationToken will be null or undefined when there are no more blobs to list?

File: src/azureStorageExplorer/blobContainers/blobContainerNode.ts [hunk 2]
```diff
@@ -115,7 +137,7 @@ export class BlobContainerNode implements IAzureParentTreeItem, ICopyUrl {
     }
 
     // This is the public entrypoint for azureStorage.uploadBlockBlob
-    public async uploadBlockBlob(node: IAzureParentNode<BlobContainerNode>): Promise<void> {
+    public async uploadBlockBlob(node: IAzureParentNode<BlobContainerNode>, output: vscode.OutputChannel): Promise<void> {
```
Comment (nturinski) [inline src/azureStorageExplorer/blobContainers/blobContainerNode.ts]: Have you seen Eric's implementation of extensionVariables?  I don't know your opinion on it, but it feels easier to utilize than passing the output channel through functions.

File: thirdpartynotices.txt [hunk 3]
```diff
@@ -8,6 +8,7 @@ are grateful to these developers for their contribution to open source.
 1. node-winreg (https://github.com/fresc81/node-winreg)
 2. fs-extra (https://github.com/jprichardson/node-fs-extra)
 3. copy-paste (https://github.com/xavi-/node-copy-paste)
+4. glob (https://github.com/isaacs/node-glob)
```
Comment (nturinski) [inline thirdpartynotices.txt]: Friendly reminder to register 3rd parties 😁

File: src/extension.ts [hunk 4]
```diff
@@ -39,34 +41,41 @@ export function activate(context: vscode.ExtensionContext): void {
...
         node.openInPortal();
-    }
-    );
-    actionHandler.registerCommand("azureStorage.configureStaticWebsite", (node: IAzureNode<IAzureTreeItem>) => {
-        if (node) {
-            let featureQuery = "feature.staticwebsites=true"; // Needed until preview is public
-            let resourceId = `${node.id}/staticWebsite`;
-            node.openInPortal(resourceId, { queryPrefix: featureQuery });
+    });
+    actionHandler.registerCommand("azureStorage.configureStaticWebsite", async (node: IAzureNode<IAzureTreeItem>) => {
```
Comment (nturinski) [inline src/extension.ts]: Why not make this its own function in a separate file?

File: src/extension.ts [hunk 5]
```diff
@@ -39,34 +41,41 @@ export function activate(context: vscode.ExtensionContext): void {
...
+        if (!node) {
+            node = <IAzureNode<StorageAccountNode>>await tree.showNodePicker(StorageAccountNode.contextValue);
+        }
+        if (node.treeItem.contextValue === BlobContainerNode.contextValue) {
+            // Currently the portal only allows configuring at the storage account level, so retrieve the storage account node
+            let storageAccountNode = node.parent && node.parent.parent;
+            console.assert(!!storageAccountNode && storageAccountNode.treeItem.contextValue === StorageAccountNode.contextValue, "Couldn't find storage account node for container");
```
Comment (nturinski) [inline src/extension.ts]: What's the difference between using a console.assert vs. throwing an error?

Also, is there any concern about storageAccountNode.treeItem being undefined and then checking its contextValue?

File: src/constants.ts [hunk 6]
```diff
@@ -0,0 +1,12 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+export const extensionPrefix: string = 'appService';
```
Comment (nturinski) [inline src/constants.ts]: Is the prefix being appService intentional?

File: src/azureStorageExplorer/storageAccounts/storageAccountNode.ts [hunk 7]
```diff
@@ -23,19 +25,35 @@ export class StorageAccountNode implements IAzureParentTreeItem {
 
     public id: string = this.storageAccount.id;
     public label: string = this.storageAccount.name;
-    public contextValue: string = 'azureStorageAccount';
+    public static contextValue: string = 'azureStorageAccount';
+    public contextValue: string = StorageAccountNode.contextValue;
```
Comment (nturinski) [inline src/azureStorageExplorer/storageAccounts/storageAccountNode.ts]: I don't quite understand why there are two contextValue properties.

File: src/azureStorageExplorer/storageAccountProvider.ts [hunk 8]
```diff
@@ -10,6 +10,8 @@ import { IChildProvider, IAzureTreeItem, IAzureNode } from 'vscode-azureextensio
 import { StorageAccountNode } from './storageAccounts/storageAccountNode';
 
 export class StorageAccountProvider implements IChildProvider {
+    public childTypeLabel: string = "storage account";
```
Comment (nturinski) [inline src/azureStorageExplorer/storageAccountProvider.ts]: nit: Not "Storage Account"?

File: src/azureStorageExplorer/blobContainers/quickPickUtils.ts [hunk 9]
```diff
@@ -0,0 +1,56 @@
+/*---------------------------------------------------------------------------------------------
```
Comment (nturinski) [inline src/azureStorageExplorer/blobContainers/quickPickUtils.ts]: This seems like a lot of reused code from the ui package. Did we have to make a new util file for the workspaceFolders?

File: src/azureStorageExplorer/blobContainers/blobContainerNode.ts [hunk 10]
```diff
@@ -295,4 +522,14 @@ export class BlobContainerNode implements IAzureParentTreeItem, ICopyUrl {
             });
         });
     }
+
+    private throwIfCanceled(cancellationToken: vscode.CancellationToken | undefined, properties: TelemetryProperties, cancelStep: string): void {
```
Comment (nturinski) [inline src/azureStorageExplorer/blobContainers/blobContainerNode.ts]: The ui package should handle cancellations in this way.

File: src/azureStorageExplorer/blobContainers/blobContainerNode.ts [hunk 11]
```diff
@@ -181,32 +203,237 @@ export class BlobContainerNode implements IAzureParentTreeItem, ICopyUrl {
...
+    private async uploadFileToBlockBlob(filePath: string, blobPath: string, output: vscode.OutputChannel): Promise<void> {
+        let blobFriendlyPath = `${this.friendlyContainerName}${blobPath}`;
+        if (output) {
+            // output.show();
+            output.appendLine(`Uploading ${filePath} as ${blobFriendlyPath}`);
+        }
         const blobService = azureStorage.createBlobService(this.storageAccount.name, this.key.value);
         let speedSummary;
-        const promise = new Promise((resolve, reject) => {
+        const uploadPromise = new Promise((resolve, reject) => {
```
Comment (nturinski) [inline src/azureStorageExplorer/blobContainers/blobContainerNode.ts]: Is this only used if there is no output?

File: src/azureStorageExplorer/blobContainers/blobContainerNode.ts [hunk 12]
```diff
@@ -181,32 +203,237 @@ export class BlobContainerNode implements IAzureParentTreeItem, ICopyUrl {
...
     }
 
-    private async uploadFileToBlockBlob(filePath: string, blobPath: string): Promise<void> {
-        let blobFullDisplayPath = `${this.storageAccount.name}/${this.container.name}/${blobPath}`;
-        channel.show();
-        channel.appendLine(`Uploading ${filePath} as ${blobFullDisplayPath}`);
+    private async uploadFileToBlockBlob(filePath: string, blobPath: string, output: vscode.OutputChannel): Promise<void> {
+        let blobFriendlyPath = `${this.friendlyContainerName}${blobPath}`;
+        if (output) {
+            // output.show();
```
Comment (nturinski) [inline src/azureStorageExplorer/blobContainers/blobContainerNode.ts]: nit: Remove comment

File: package.json [hunk 13]
```diff
@@ -200,10 +212,22 @@
...
+				{
+					"command": "azureStorage.deployStaticWebsite",
+					"when": "view == azureStorage && config.azureStorage.preview.staticWebsites",
+					"group": "navigation@2"
+				},
 				{
 					"command": "azureStorage.refresh",
 					"when": "view == azureStorage",
-					"group": "navigation"
+					"group": "navigation@99"
```
[the remark this answers is not in the document]
Comment (StephenWeatherford) [reply package.json]: Yeppers

File: src/azureStorageExplorer/blobContainers/blobContainerNode.ts (hunk 1 above)
Comment (StephenWeatherford) [reply src/azureStorageExplorer/blobContainers/blobContainerNode.ts]: correct

File: thirdpartynotices.txt (hunk 3 above)
Comment (StephenWeatherford) [reply thirdpartynotices.txt]: Already done, but thanks for reminder.

File: src/azureStorageExplorer/blobContainers/blobContainerNode.ts (hunk 2 above)
Comment (StephenWeatherford) [reply src/azureStorageExplorer/blobContainers/blobContainerNode.ts]: Might move to that later.

File: src/azureStorageExplorer/storageAccountProvider.ts (hunk 8 above)
Comment (StephenWeatherford) [reply src/azureStorageExplorer/storageAccountProvider.ts]: Personally I think these should all be sentence case, not title case, but I see the others are all title case, so I guess I'll change mine to be consistent.

File: src/azureStorageExplorer/blobContainers/quickPickUtils.ts (hunk 9 above)
Comment (StephenWeatherford) [reply src/azureStorageExplorer/blobContainers/quickPickUtils.ts]: I have a work item to merge with common.

File: src/azureStorageExplorer/blobContainers/blobContainerNode.ts (hunk 10 above)
Comment (StephenWeatherford) [reply src/azureStorageExplorer/blobContainers/blobContainerNode.ts]: This is dealing with the cancel button in the window.withProgress dialog, which has to be polled.  Or am I misunderstanding you?

File: src/azureStorageExplorer/blobContainers/blobContainerNode.ts (hunk 11 above)
Comment (StephenWeatherford) [reply src/azureStorageExplorer/blobContainers/blobContainerNode.ts]: No.

        if (output) {
            await awaitWithProgress(`Uploading ${blobPath}`, channel, *uploadPromise*, () => {

Reformatted to make more obvious.

File: src/azureStorageExplorer/storageAccounts/storageAccountActionHandlers.ts [hunk 14]
```diff
@@ -29,3 +38,76 @@ async function copyConnectionString(node: IAzureNode<StorageAccountNode>): Promi
...
+
+        if (destNode) {
+            let contextValue = destNode.treeItem.contextValue;
+            if (contextValue === StorageAccountNode.contextValue) {
+                // Called from storage account node
+                destAccountNode = <IAzureParentNode<StorageAccountNode>>destNode;
+            } else if (contextValue === BlobContainerNode.contextValue) {
+                // Called from blob container node
+                destContainerNode = <IAzureParentNode<BlobContainerNode>>destNode;
+            }
```
[the remark this answers is not in the document]
Comment (StephenWeatherford) [reply src/azureStorageExplorer/storageAccounts/storageAccountActionHandlers.ts]: It's assigning to different variables:

            if (contextValue === StorageAccountNode.contextValue) {
                // Called from storage account node
                destAccountNode = <IAzureParentNode<StorageAccountNode>>destNode;
            } else if (contextValue === BlobContainerNode.contextValue) {
                // Called from blob container node
                destContainerNode = <IAzureParentNode<BlobContainerNode>>destNode;

In the latter case I have to do additional work to get to the storage account node.

File: src/azureStorageExplorer/storageAccounts/storageAccountActionHandlers.ts [hunk 15]
```diff
@@ -29,3 +38,76 @@ async function copyConnectionString(node: IAzureNode<StorageAccountNode>): Promi
...
+        }
+    }
+
+    // Ask first for destination account/container if needed since it might require configuration and don't want to have user
+    // select source location only to have to possibly cancel.
+    if (!destNode) {
+        destAccountNode = <IAzureParentNode<StorageAccountNode>>await _tree.showNodePicker(StorageAccountNode.contextValue);
+    }
+
+    if (!destContainerNode) {
```
[the remark this answers is not in the document]
Comment (StephenWeatherford) [reply src/azureStorageExplorer/storageAccounts/storageAccountActionHandlers.ts]: Make assertion more explicit:

    console.assert(!!destAccountNode || !!destContainerNode, "Should have a storage account or container node by now");

We will ask for storage account if the context is neither account nor container

File: src/extension.ts (hunk 4 above)
Comment (StephenWeatherford) [reply src/extension.ts]: Deals with both accounts and containers, and not enough code to worry about, IMHO.

File: src/constants.ts (hunk 6 above)
Comment (StephenWeatherford) [reply src/constants.ts]: Nope, but wasn't used, removed.  Will deal with it when I move files browse to common.

File: src/azureStorageExplorer/storageAccounts/storageAccountNode.ts (hunk 7 above)
Comment (StephenWeatherford) [reply src/azureStorageExplorer/storageAccounts/storageAccountNode.ts]: Once is a static that can be referenced outside this node without an instance.  The other is the instance variable that's required as part of implementing IAzureTreeItem.

File: src/extension.ts (hunk 5 above)
Comment (StephenWeatherford) [reply src/extension.ts]: If storageAccountNode.treeItem == undefined, I'd say something is seriously wrong in the tree code and it'll die a horrible death soon enough.  :-)

Comment (nturinski) [reply src/extension.ts]: Fair enough 😆 

Do we want to stifle the error with console.assert?  I would think we want to throw an error to at least telemetry.

Comment (StephenWeatherford) [reply src/extension.ts]: console.assert actually throws an error.

My reasons for using console.assert:
1) It makes it clear that I'm "asserting" something should be true
2) Easier to code/read than an if/else

File: src/azureStorageExplorer/blobContainers/blobContainerNode.ts (hunk 10 above)
Comment (nturinski) [reply src/azureStorageExplorer/blobContainers/blobContainerNode.ts]: Ah, I see.  I didn't think about the withProgress dialog.  I'll add an issue to add that to our ui package.

**Pull request merged**